src/profilers/profile-serializer.ts

```diff
@@ -0,0 +1,145 @@
+import {perftools} from '../profile';
+import {getIndexOrAdd} from '../util';
+import {WallProfile, WallProfileNode} from '../v8-types';
+// A stack of function UIDs.
+type Stack = Array<number>;
+
+// Helper class used to convert v8 profile into profile format used by
+// Stackdriver Profiler.
+class ProfileSerializer {
+  private samples: Array<perftools.profiles.Sample>;
+  private locations: Array<perftools.profiles.Location>;
+  private functions: Array<perftools.profiles.Function>;
+  private locationMap: Map<number, perftools.profiles.Location>;
+  private functionMap: Map<number, perftools.profiles.Function>;
+  private strings: string[];
+  private samplingInterval: number;  // microseconds, defaults to 1000us
+
+  constructor() {
+    this.reset();
+  }
+
+  // Converts profile returned by v8 wall profiler to profile with format used
+  // by Stackdriver Profilers.
+  serializeWallProfile(prof: WallProfile, sampleInterval: number):
+      perftools.profiles.IProfile {
+    this.reset();
+    this.samplingInterval = sampleInterval;
+
+    let sampleValueType = this.sampleValueType();
+    let timeValueType = this.timeValueType();
+
+    this.serializeNode(prof.topDownRoot, []);
+
+    return {
+      sampleType: [sampleValueType, timeValueType],
+      sample: this.samples,
+      location: this.locations,
+      function: this.functions,
+      stringTable: this.strings,
+      // opt drop_frames
+      // opt keep_frames
+      timeNanos: 1000 * 1000 * prof.endTime,                         // Nanos
+      durationNanos: 1000 * 1000 * (prof.endTime - prof.startTime),  // Nanos
+
+      periodType: timeValueType,
+      period: this.samplingInterval
+    };
+  }
+
+  // clears fields used to build a profile.
+  private reset() {
+    this.samples = [];
+    this.locations = [];
+    this.functions = [];
+    this.locationMap = new Map();
+    this.functionMap = new Map<number, perftools.profiles.Function>();
+    this.strings = [''];
+    this.samplingInterval = 1000;
+  }
+
+  /* Adds samples from a node and it's children to the fields tracking
+   * profile serialization.
+   *
+   * node - the node which is serialized
+   * stack - the stack trace to the current node.
+   */
+  private serializeNode(node: WallProfileNode, stack: Stack) {
+    let that = this;
```

Comment:
Change the `function(child)` below to a fat arrow function (`(child) => { ... }`), and this would not be needed anymore. Arrow functions keep the lexical this.

Even better, use a for..of loop instead of the `forEach`. `for (let child of node.children) {... }`. This avoids the extra closure allocation / function dispatch per iteration.